sklearn/discriminant_analysis.py

```diff
@@ -806,14 +806,17 @@ def fit(self, X, y):
             Xgc = Xg - meang
             # Xgc = U * S * V.T
             _, S, Vt = np.linalg.svd(Xgc, full_matrices=False)
-            rank = np.sum(S > self.tol)
-            if rank < n_features:
-                warnings.warn("Variables are collinear")
             S2 = (S ** 2) / (len(Xg) - 1)
             S2 = ((1 - self.reg_param) * S2) + self.reg_param
+            cov_reg = np.dot(S2 * Vt.T, Vt)
+            det = linalg.det(cov_reg)
```

Comment:
Computing the determinant is expansive, let's just compute `rank = np.sum(S > self.tol)` as before (which is cheaper once we have the result of the SVD) and raise the exception when `rank < n_features`.